### PR TITLE
sys/net/routing/rpl: use types from `byteorder.h` for `uint16_t`

### PR DESCRIPTION
**Pull request by BytesGalore**
Rationale:
The byteorder has not been considered when receiving DIOs, which resulted in wrong values when trying to join a DODAG.
This PR replaces the former `uint16_t` variables  with the bytorder type [`network_uint16_t`](https://github.com/RIOT-OS/RIOT/blob/master/core/include/byteorder.h#L110) from [1] and enforces the correct byteorder.
This PR is an alternative to #2431.

[1] https://github.com/RIOT-OS/RIOT/blob/master/core/include/byteorder.h

### CODE REVIEW

File: sys/net/routing/rpl/rpl.c [hunk 1]
```diff
@@ -370,24 +370,24 @@ void _rpl_update_routing_table(void) {
 
         for (uint8_t i = 0; i < rpl_max_routing_entries; i++) {
             if (rt[i].used) {
-                if (rt[i].lifetime <= 1) {
+                if (NTOHS(rt[i].lifetime.u16) <= 1) {
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl.c]: shouldn't the lifetime value be in host byte order at this point anyway?

File: sys/net/routing/rpl/rpl.c [hunk 2]
```diff
@@ -482,7 +482,7 @@ void rpl_add_routing_entry(ipv6_addr_t *addr, ipv6_addr_t *next_hop, uint16_t li
     rpl_routing_entry_t *entry = rpl_find_routing_entry(addr);
 
     if (entry != NULL) {
-        entry->lifetime = lifetime;
+        entry->lifetime.u16 = HTONS(lifetime);
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl.c]: I c, the lifetime value is stored in network byte order in the routing table

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl.c]: yep

File: examples/rpl_udp/rpl.c [hunk 3]
```diff
@@ -167,7 +167,7 @@ void rpl_udp_dodag(int argc, char **argv)
     printf("Part of Dodag:\n");
     printf("%s\n", ipv6_addr_to_str(addr_str, IPV6_MAX_ADDR_STR_LEN,
                                     (&mydodag->dodag_id)));
-    printf("my rank: %d\n", mydodag->my_rank);
+    printf("my rank: %d\n", NTOHS(mydodag->my_rank.u16));
```
Comment (Kijewski) [inline examples/rpl_udp/rpl.c]: byteorder_ntohs(mydodag->my_rank)
No member access to .u16, .u32 etc should ever be needed.

Comment (BytesGalore) [inline examples/rpl_udp/rpl.c]: But when I remove the member access and the conversion to host byteorder I get:
```
RIOT/examples/rpl_udp/rpl.c:170:5: warning: format '%d' expects argument of type 'int', but argument 2 has type 'network_uint16_t' [-Wformat=]
     printf("my rank: %d\n", (mydodag->my_rank));
     ^
```
On:
`gcc version 4.9.3 20141119 (release) [ARM/embedded-4_9-branch revision 218278]`
for `iot-lab_M3`

Comment (Kijewski) [inline examples/rpl_udp/rpl.c]: You omitted the byteorder_ntohs :)

Comment (BytesGalore) [inline examples/rpl_udp/rpl.c]: uhh, ok we all know I need :eyeglasses:

**Comment by cgundogan** [comment]
I actually would expect that you shouldn't need byte order switching functions when accessing the values within the same host. So, I would prefer to only use byte ordering when receiving / transmitting packets. Consequently, we should consider to only change `uint16_t` => `network_uint16_t` in structs that are used for communication (identified by `__attribute__((packed))`) like `rpl_dio_t`, and not e.g. `rpl_dodag_t`

**Comment by BytesGalore** [comment]
now uses the proper conversion functions

**Comment by BytesGalore** [comment]
@cgundogan basically I'm in your opinion, but when we switch consequently to `network_uint16_t` there is no room for conversion mismatch anymore

**Comment by authmillenon** [comment]
> but when we switch consequently to network_uint16_t there is no room for conversion mismatch anymore

This looks a little bit overkill to me. `network_uint16_t` should only be used for integers that are actually send over the network (members of header definitions e.g.). You are constantly converting the values otherwise.

**Comment by authmillenon** [comment]
Additionally, this is the main problem in the network stack, why there are so many byteorder problems: Every layer feels responsible to HTONS/NTOHS everything they get and no one knows at anytime for sure which order they are in.

**Comment by BytesGalore** [comment]
Hmm, then we can just close this PR and use #2431, which only takes care of the sent/received variables that are actually used.

**Comment by authmillenon** [comment]
:+1:

**Comment by BytesGalore** [comment]
close in favour of #2431

**Pull request closed without merging**